### PR TITLE
Add Result and ExperimentResult structs

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Start of refactor of simulator result data types.  This adds C++ structs for simulator results rather than using pure json internally. The goal is that these classes can be wrapped into python using pybind or cython to directly build a python result object without going via json/str conversion for local simulation.

### Details and comments

* Adds `Result` class
* Adds `ExperimentResult` class
* Renames `OutputData` to `ExperimentData`

### CODE REVIEW

File: src/framework/results/experiment_result.hpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+#include "framework/results/experiment_data.hpp"
+
+namespace AER {
+
+//============================================================================
+// Result container for Qiskit-Aer
+//============================================================================
+
+struct ExperimentResult {
```
Comment (atilag) [inline src/framework/results/experiment_result.hpp]: This interface is very similar to `Result` one, Didn't want to inherit from it?

File: src/framework/results/experiment_result.hpp [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  // Serialize engine data to JSON
+  json_t json() const;
+};
+
+
+//------------------------------------------------------------------------------
+// Add metadata
+//------------------------------------------------------------------------------
+template <typename T>
+void ExperimentResult::add_metadata(const std::string &key, const T &meta) {
```
Comment (atilag) [inline src/framework/results/experiment_result.hpp]: This is exactly the same method found in `Result::add_metadata()`... inheriting from 
`Result` will remove the need to duplicate code.

File: src/framework/results/experiment_result.hpp (hunk 1 above)
Comment (chriseclectic) [reply src/framework/results/experiment_result.hpp]: `Result` contains `ExperimentResult` objects as members though

Comment (atilag) [reply src/framework/results/experiment_result.hpp]: Inherit by composition, perfect.

**Pull request merged**